presto-verifier/src/test/java/com/facebook/presto/verifier/rewrite/TestQueryRewriter.java

```diff
@@ -294,7 +298,163 @@ public void testRewriteNonStorableStructuredTypes()
                         "    CAST(ROW (NULL) AS ROW(BIGINT))");
     }
 
-    private void assertShadowed(
+    @Test
+    public void constructMakeFunctionCallSubstituteMap()
+    {
+        assertEqualsDeep(FunctionCallRewriter.constructFunctionCallSubstituteMap("/func1(a,b)/func2(b)/"),
+                ImmutableMap.builder().put(QualifiedName.of("func1"),
+                        new FunctionCallRewriter.FunctionCallSubstitute(QualifiedName.of("func2"), ImmutableList.of(1))).build());
+        assertEqualsDeep(FunctionCallRewriter.constructFunctionCallSubstituteMap("/func1(,,1c)/func2(1c)/"),
+                ImmutableMap.builder().put(QualifiedName.of("func1"),
+                        new FunctionCallRewriter.FunctionCallSubstitute(QualifiedName.of("func2"), ImmutableList.of(2))).build());
+        assertEqualsDeep(FunctionCallRewriter.constructFunctionCallSubstituteMap("/func1(a,_,b)/func2(_)/"),
+                ImmutableMap.builder().put(QualifiedName.of("func1"),
+                        new FunctionCallRewriter.FunctionCallSubstitute(QualifiedName.of("func2"), ImmutableList.of(1))).build());
+        assertEqualsDeep(FunctionCallRewriter.constructFunctionCallSubstituteMap("/func1(a,b)/func2(b)/,/func3(a,)/func4(a)/"),
+                ImmutableMap.builder().put(QualifiedName.of("func1"),
+                        new FunctionCallRewriter.FunctionCallSubstitute(QualifiedName.of("func2"), ImmutableList.of(1))).put(QualifiedName.of("func3"),
+                        new FunctionCallRewriter.FunctionCallSubstitute(QualifiedName.of("func4"), ImmutableList.of(0))).build());
+
+        assertTrue(FunctionCallRewriter.constructFunctionCallSubstituteMap("").isEmpty());
+        assertTrue(FunctionCallRewriter.constructFunctionCallSubstituteMap("/func1(a,b)//").isEmpty());
+    }
+
+    @Test
+    public void testRewriteFunctionCalls()
+    {
+        VerifierConfig verifierConfig = new VerifierConfig().setNonDeterministicFunctionSubstitutes(
+                "/approx_distinct(x)/count(x)/," +
+                        "/approx_percentile(x,_)/avg(x)/," +
+                        "/arbitrary(x)/min(x)/," +
+                        "/map_agg(x,)/max(x)/," +
```

Comment:
Is it valid to replace map_agg with max? Would you clarify?

Reply:
Still wondering about this.

Reply:
Probably not.
Would we support something like
`"/map_agg(x,y)/map_sort(map_agg(x,y))/,"`
?

Reply:
> Is it valid to replace map_agg with max? Would you clarify?
It's illegal. Removed this test case.

Reply:
> Probably not. Would we support something like `"/map_agg(x,y)/map_sort(map_agg(x,y))/,"` ?

With current logic, this or nested function call is not supported. But I'm thinking of iterating on this part with one more PR, with more sophisticated parsing and internal structure. The use case of map_agg and array_agg are noted and need to think over.

Reply:
@spershin @gggrace14 Sergey, Ge, FYI, array_sort requires that array elements are orderable, which is not the case of map, e.g. applying array_sort for array(map(K,v)) would fail.